File: go/linkedlist/linked_list.go
package linkedlist

import (
	"fmt"
	"errors"
)

type LinkedList struct {
	dummyHead Node
	size int
}

func New() *LinkedList {
	// 注意只有上面的 dummyHead 声明成结构体的时候才可以这样初始化
	// 如果是指向 Node 的指针的话，就必须显示添加一个 Node 了
	return &LinkedList{}
}

func (l *LinkedList) GetSize() int {
	return l.size
}

func (l *LinkedList) IsEmpty() bool {
	return l.size == 0
}

func (l *LinkedList) Insert(index int, elem interface{}) error {
	if index < 0 || index > l.size {
		return errors.New("Index out of range")
	}

	prev := &l.dummyHead
	for i := 0; i < index; i++ {
		prev = prev.next
	}
	node := Node{data: elem,}
	node.next = prev.next
	prev.next = &node
	l.size++
	return nil
}

func (l *LinkedList) AddFirst(elem interface{}) {
	l.Insert(0, elem)
}

func (l *LinkedList) AddLast(elem interface{}) {
	l.Insert(l.size, elem)
}

func (l *LinkedList) Remove(index int) (elem interface{}, err error) {
	if index < 0 || index > l.size - 1 {
		err = errors.New("Index out of range")
		return
	}

	prev := &l.dummyHead
	for i := 0; i < index; i++ {
		prev = prev.next
	}
	node := prev.next
	prev.next = node.next
	node.next = nil
	elem = node.data
	l.size--
	return
}

func (l *LinkedList) RemoveFirst() (interface{}, error) {
	return l.Remove(0)
}

func (l *LinkedList) RemoveLast() (interface{}, error) {
	return l.Remove(l.size - 1)
}

func (l *LinkedList) Get(index int) (elem interface{}, err error) {
	if index < 0 || index > l.size {
		err = errors.New("Index out of range")
		return
	}

	current := l.dummyHead.next
	for i := 0; i < index; i++ {
		current = current.next
	}
	elem = current.data
	return
}

func (l *LinkedList) GetFirst() (interface{}, error) {
	return l.Get(0)
}

func (l *LinkedList) GetLast() (interface{}, error) {
	return l.Get(l.size - 1)
}

func (l *LinkedList) Set(index int, elem interface{}) error {
	if index < 0 || index > l.size {
		return errors.New("Index out of range")
	}

	current := l.dummyHead.next
	for i := 0; i < index; i++ {
		current = current.next
	}
	current.data = elem
	return nil
}

func (l *LinkedList) Contains(elem interface{}) bool {
	current := l.dummyHead.next
	for i := 0; i < l.size; i++ {
		if current.data == elem {
			return true
		} else {
			current = current.next
		}
	}
	return false
}

func (l *LinkedList) String() string {
	str := "LinkedList: "
	current := l.dummyHead.next
	for i := 0; i < l.size; i++ {
		str += fmt.Sprint(current.data) + " -> "
		current = current.next
	}

	str += "Null"
	return str
}

// 补充内容
// 统计列表中指定元素出现的次数
func (l *LinkedList) Count(elem interface{}) int {
	count := 0
	for cur := l.dummyHead.next; cur != nil; cur = cur.next {
		if cur.data == elem {
			count++
		}
	}
	return count
}

// 拷贝一份链表
func (l *LinkedList) Copy() *LinkedList {
	newList := New()

	for cur := l.dummyHead.next; cur != nil; cur = cur.next {
		newList.AddLast(cur.data)
	}

	return newList
}

// 优化后的拷贝操作
// 理论上来说复杂度为 O(n) 级别
func (l *LinkedList) OptimizedCopy() *LinkedList {
	nodes := []*Node{}
	newList := New()

	for cur := l.dummyHead.next; cur != nil; cur = cur.next {
		node := &Node{data: cur.data,}
		nodes = append(nodes, node)
	}

	for i := 0; i < len(nodes) - 1; i++ {
		nodes[i].next = nodes[i+1]
	}


	newList.size = l.GetSize()
	if len(nodes) == 0 {
		return newList
	}
	head := Node{data: nil, next: nodes[0],}
	newList.dummyHead = head
	return newList
}

File: js/LinkedList/LinkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.dummyHead = new Node(null);
    this.size = 0;
  }

  getSize() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }

  insert(index, data) {
    if (index < 0 || index > this.size) {
      throw new Error("Index out of the range");
    }

    let prev = this.dummyHead;
    // 这样想：遍历完索引 i 后，prev 指向索引 i 的节点，因此当 i = index 退出循环时
    // i 指向 index - 1 的节点
    for(let i = 0; i < index; i++) {
      prev = prev.next;
    }

    const node = new Node(data);
    node.next = prev.next;
    prev.next = node;
    this.size++;
  }

  addFirst(data) {
    this.insert(0, data);
  }

  addLast(data) {
    this.insert(this.size, data);
  }

  remove(index) {
    if (index < 0 || index > this.size - 1) {
      throw new Error("Index out of the range");
    }

    let prev = this.dummyHead;
    for (let i = 0; i < index; i++) {
      prev = prev.next;
    }

    let node = prev.next;
    prev.next = node.next;
    const data = node.data;
    // 这里将 node 设置为 null 没什么作用吧，并且只是一个引用
    node.next = null;
    this.size--;
    return data;
  }

  removeFirst() {
    return this.remove(0);
  }

  removeLast() {
    return this.remove(this.size - 1);
  }

  get(index) {
    if (index < 0 || index >= this.size) {
      throw new Error("Index out of the range");
    }

    let current = this.dummyHead.next;
    // 这里就不一样了，当遍历完索引 i 后, current 指向了索引为 i + 1 元素
    // 所以当循环终止时，current 指向了索引为 index 的节点
    for (let i = 0; i < index; i++) {
      current = current.next;
    }

    return current.data;
  }

  getFirst() {
    return this.get(0);
  }

  getLast() {
    return this.get(this.size-1)
  }

  set(index, data) {
    if (index < 0 || index >= this.size) {
      throw new Error("Index out of the range");
    }

    let current = this.dummyHead;
    for (let i = 0; i < index; i++) {
      current = current.next;
    }

    current.data = data;
  }

  toString() {
    let str = `LinkedList: Size = ${this.size}\n`;
    let current = this.dummyHead.next;
    while(current != null) {
      str += `${current.data} -> `;
      current = current.next;
    }
    str += 'Null';
    return str;
  }
}

module.exports = LinkedList;
